File: internal/domain/edge.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// Edge represents a directed knowledge-graph edge between two Pages.
type Edge struct {
	ID           uuid.UUID      `json:"id"`
	SourcePageID uuid.UUID      `json:"source_page_id"`
	TargetPageID uuid.UUID      `json:"target_page_id"`
	EdgeType     string         `json:"edge_type"`
	Metadata     map[string]any `json:"metadata,omitempty"`
	CreatedBy    uuid.UUID      `json:"created_by"`
	CreatedAt    time.Time      `json:"created_at"`
}

// CreateEdgeRequest carries the fields needed to create a new Edge.
type CreateEdgeRequest struct {
	TargetPageID uuid.UUID      `json:"target_page_id"`
	EdgeType     string         `json:"edge_type"`
	Metadata     map[string]any `json:"metadata,omitempty"`
}

// GraphNeighbors is the response for the GET /pages/:id/graph endpoint.
type GraphNeighbors struct {
	Center    GraphNode   `json:"center"`
	Edges     []GraphEdge `json:"edges"`
	NodeCount int         `json:"node_count"`
	EdgeCount int         `json:"edge_count"`
}

// GraphNode is a minimal page node used in graph responses.
type GraphNode struct {
	ID              string `json:"id"`
	Title           string `json:"title"`
	FreshnessStatus string `json:"freshness_status"`
}

// GraphEdge is an edge in a graph response.
type GraphEdge struct {
	ID        string         `json:"id"`
	Source    GraphNode      `json:"source"`
	Target    GraphNode      `json:"target"`
	EdgeType  string         `json:"edge_type"`
	Metadata  map[string]any `json:"metadata,omitempty"`
	CreatedAt string         `json:"created_at"`
}

// GraphExploreResult is the response for GET /graph/explore.
type GraphExploreResult struct {
	Nodes      []ExploreNode `json:"nodes"`
	Edges      []ExploreEdge `json:"edges"`
	TotalNodes int           `json:"total_nodes"`
	TotalEdges int           `json:"total_edges"`
	Truncated  bool          `json:"truncated"`
}

// ExploreNode is a node in the exploration subgraph.
type ExploreNode struct {
	ID              string `json:"id"`
	Title           string `json:"title"`
	SpaceID         string `json:"space_id"`
	FreshnessStatus string `json:"freshness_status"`
	ConnectionCount int    `json:"connection_count"`
	DepthFromRoot   int    `json:"depth_from_root"`
}

// ExploreEdge is an edge in the exploration subgraph.
type ExploreEdge struct {
	SourceID string `json:"source_id"`
	TargetID string `json:"target_id"`
	EdgeType string `json:"edge_type"`
}

File: web/lib/models/edge.dart
class GraphNode {
  final String id;
  final String title;
  final String? spaceId;
  final String freshnessStatus;
  final int connectionCount;
  final int depthFromRoot;

  const GraphNode({
    required this.id,
    required this.title,
    this.spaceId,
    required this.freshnessStatus,
    required this.connectionCount,
    required this.depthFromRoot,
  });

  factory GraphNode.fromJson(Map<String, dynamic> json) {
    return GraphNode(
      id: json['id'] as String,
      title: json['title'] as String,
      spaceId: json['space_id'] as String?,
      freshnessStatus: json['freshness_status'] as String? ?? 'fresh',
      connectionCount: json['connection_count'] as int? ?? 0,
      depthFromRoot: json['depth_from_root'] as int? ?? 0,
    );
  }
}

class GraphEdge {
  final String? id;
  final String sourceId;
  final String targetId;
  final String edgeType;
  final Map<String, dynamic>? metadata;
  final DateTime? createdAt;

  const GraphEdge({
    this.id,
    required this.sourceId,
    required this.targetId,
    required this.edgeType,
    this.metadata,
    this.createdAt,
  });

  factory GraphEdge.fromJson(Map<String, dynamic> json) {
    return GraphEdge(
      id: json['id'] as String?,
      sourceId: (json['source_id'] ?? json['source_page_id'] ?? '') as String,
      targetId: (json['target_id'] ?? json['target_page_id'] ?? '') as String,
      edgeType: json['edge_type'] as String? ?? 'reference',
      metadata: json['metadata'] as Map<String, dynamic>?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : null,
    );
  }

  factory GraphEdge.fromNeighborJson(Map<String, dynamic> json) {
    final source = json['source'] as Map<String, dynamic>?;
    final target = json['target'] as Map<String, dynamic>?;
    return GraphEdge(
      id: json['id'] as String?,
      sourceId: source?['id'] as String? ?? '',
      targetId: target?['id'] as String? ?? '',
      edgeType: json['edge_type'] as String? ?? 'reference',
      metadata: json['metadata'] as Map<String, dynamic>?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : null,
    );
  }
}

class GraphData {
  final List<GraphNode> nodes;
  final List<GraphEdge> edges;
  final int totalNodes;
  final int totalEdges;
  final bool truncated;

  const GraphData({
    required this.nodes,
    required this.edges,
    required this.totalNodes,
    required this.totalEdges,
    required this.truncated,
  });

  factory GraphData.fromJson(Map<String, dynamic> json) {
    final data = json['data'] as Map<String, dynamic>;
    return GraphData(
      nodes: (data['nodes'] as List? ?? [])
          .map((n) => GraphNode.fromJson(n as Map<String, dynamic>))
          .toList(),
      edges: (data['edges'] as List? ?? [])
          .map((e) => GraphEdge.fromJson(e as Map<String, dynamic>))
          .toList(),
      totalNodes: data['total_nodes'] as int? ?? 0,
      totalEdges: data['total_edges'] as int? ?? 0,
      truncated: data['truncated'] as bool? ?? false,
    );
  }
}
